File: mobile/dominio/operarios/validador-operario.ts
/**
 * Validador puro del módulo OPERARIOS — reglas multi-tenant.
 *
 * Las funciones aquí retornan `boolean` para mantener el validador puro y
 * testeable sin dependencia de strings de UI. La capa de presentación
 * (Fase 5) es responsable de mapear `false` → mensaje al usuario.
 *
 * Reglas cubiertas:
 *   - 2.2 `idPrestadorRequeridoValido` — operario multi-tenant debe tener
 *     `id_prestador > 0` (columna agregada por migration 016, FK a prestador).
 *   - 2.3 `passwordCumpleMinima` — política de longitud mínima para la
 *     contraseña cruda que el operario envía al login.
 *
 * Por ahora el módulo solo contiene reglas del change
 * `setup-inicial-multi-tenant-auth`. Las reglas históricas (cedula, email,
 * password_hash) viven en `operarios.ts` factory a través de
 * `MENSAJES_ERROR_OPERARIO`.
 */

const PASSWORD_LONGITUD_MINIMA = 8;

/**
 * Valida que un operario pertenezca a un prestador multi-tenant.
 *
 * El modelo multi-tenant exige `id_prestador > 0` (entero positivo) para
 * todo operario creado post-migration 016. El legacy `id_prestador = 0`
 * está reservado para datos importados previos al change.
 *
 * La función también rechaza decimales y `NaN`: aunque la columna SQLite
 * es INTEGER, la capa de UI podría enviar `parseFloat(input.value)` por
 * error; la guarda evita que un no-entero pase al repositorio.
 *
 * @param idPrestador valor numérico a validar.
 * @returns `true` solo si es entero y `> 0`. `false` para `0`, negativos,
 *          decimales y `NaN`.
 */
export function idPrestadorRequeridoValido(idPrestador: number): boolean {
  return Number.isInteger(idPrestador) && idPrestador > 0;
}

/**
 * Valida que la contraseña cruda tenga la longitud mínima aceptable.
 *
 * Política: 8 caracteres mínimo (alineado con NIST SP 800-63B como
 * referencia base; el hasher real y la sal se gestionan en la capa de
 * infraestructura — el dominio solo recibe la contraseña cruda desde la
 * pantalla de Login para verificar la regla de UX antes de enviar al
 * puerto de auth real en Fase 4).
 *
 * El parámetro se declara `string` pero la guarda `typeof === 'string'`
 * hace que la función sea segura ante entradas `undefined`/null que un
 * form mal inicializado podría entregar.
 *
 * @param password cadena cruda del input del usuario.
 * @returns `true` solo si es string con `length >= 8`. `false` para
 *          vacío, no-string o longitud insuficiente.
 */
export function passwordCumpleMinima(password: string): boolean {
  return typeof password === 'string' && password.length >= PASSWORD_LONGITUD_MINIMA;
}

File: backend/src/MediApp.Api/Dominio/Validacion/ValidadorOperario.cs
using System.Text.RegularExpressions;

namespace MediApp.Api.Dominio.Validacion;

/// <summary>
/// Validador puro del módulo OPERARIOS — reglas multi-tenant.
///
/// Espejo simétrico del validador mobile
/// <c>mobile/dominio/operarios/validador-operario.ts</c>. Las funciones
/// retornan <c>bool</c> para mantener el validador testeable sin acoplar a
/// strings de UI ni a FluentValidation.
///
/// Aplicar ANTES de persistir:
///   - <c>EsCedulaValida</c> → 6-12 dígitos (<c>^\d{6,12}$</c>).
///   - <c>EsNombreValido</c> → 1-150 caracteres.
///   - <c>EsEmailValido</c> → formato email RFC light; <c>null</c> o vacío
///     rechazados (la columna es <c>NOT NULL</c> en DB).
///   - <c>EsPasswordValida</c> → ≥ 8 caracteres crudos (espejo de
///     <c>passwordCumpleMinima</c> del mobile; el hash se valida aparte).
///   - <c>EsIdPrestadorValido</c> → entero > 0 (espejo de
///     <c>idPrestadorRequeridoValido</c>; <c>0</c> está reservado para
///     prestador legacy "EPC-LEGACY").
///   - <c>EsRolValido</c> → ∈ {operario, supervisor, admin}.
///   - <c>EsEstadoValido</c> → ∈ {activo, inactivo}.
/// </summary>
public static class ValidadorOperario
{
    private const int LONGITUD_NOMBRE_MAXIMA = 150;
    private const int PASSWORD_LONGITUD_MINIMA = 8;

    private static readonly string[] RolesValidos = { "operario", "supervisor", "admin" };
    private static readonly string[] EstadosValidos = { "activo", "inactivo" };

    private static readonly Regex CedulaRegex =
        new(@"^\d{6,12}$", RegexOptions.Compiled);

    private static readonly Regex EmailRegex =
        new(@"^[^@\s]+@[^@\s]+\.[^@\s]+$", RegexOptions.Compiled);

    /// <summary>
    /// Cédula del operario (6-12 dígitos). Mismo regex que mobile
    /// <c>REGEX_CEDULA = /^\d{6,12}$/</c> y backend <c>OperarioValidator</c>.
    /// </summary>
    public static bool EsCedulaValida(string? cedula)
    {
        return cedula is not null && CedulaRegex.IsMatch(cedula);
    }

    /// <summary>1-150 caracteres. La columna <c>nombre</c> es NOT NULL.</summary>
    public static bool EsNombreValido(string? nombre)
    {
        if (string.IsNullOrEmpty(nombre)) return false;
        return nombre.Length <= LONGITUD_NOMBRE_MAXIMA;
    }

    /// <summary>
    /// Email con formato RFC light (usuario@dominio.tld). Null o vacío
    /// rechazado: la columna <c>email</c> es NOT NULL en DB.
    /// </summary>
    public static bool EsEmailValido(string? email)
    {
        return email is not null && EmailRegex.IsMatch(email);
    }

    /// <summary>
    /// Contraseña cruda con longitud mínima de 8 caracteres. Espejo del
    /// validador mobile <c>passwordCumpleMinima</c>.
    ///
    /// Decisión: la validación opera sobre la contraseña CRUDA que llega
    /// del formulario. El hash (<c>PasswordHash</c>) se valida aparte en
    /// la columna <c>Operario.PasswordHash</c> (max 200 chars, NOT NULL).
    /// </summary>
    public static bool EsPasswordValida(string? password)
    {
        return password is not null && password.Length >= PASSWORD_LONGITUD_MINIMA;
    }

    /// <summary>
    /// <c>id_prestador</c> &gt; 0 (entero positivo). Espejo del validador
    /// mobile <c>idPrestadorRequeridoValido</c>. <c>0</c> está reservado
    /// para el prestador legacy "EPC-LEGACY" (ver migration 009).
    /// </summary>
    public static bool EsIdPrestadorValido(int idPrestador)
    {
        return idPrestador > 0;
    }

    /// <summary>Rol ∈ {operario, supervisor, admin}. Case-sensitive.</summary>
    public static bool EsRolValido(string? rol)
    {
        return rol is not null && Array.IndexOf(RolesValidos, rol) >= 0;
    }

    /// <summary>Estado ∈ {activo, inactivo}. Case-sensitive.</summary>
    public static bool EsEstadoValido(string? estado)
    {
        return estado is not null && Array.IndexOf(EstadosValidos, estado) >= 0;
    }
}
